File: java-app/src/main/java/com/showcase/java/domain/store/EventStore.java
package com.showcase.java.domain.store;

import com.showcase.java.domain.event.DomainEvent;

import java.util.List;

public interface EventStore {
    void append(String aggregateId, List<DomainEvent> events);
    List<DomainEvent> load(String aggregateId);
    List<DomainEvent> loadAll();
}

File: kotlin-app/src/main/kotlin/com/showcase/kotlin/domain/store/EventStore.kt
package com.showcase.kotlin.domain.store

import com.showcase.kotlin.domain.event.DomainEvent
import java.util.concurrent.ConcurrentHashMap

interface EventStore {
    suspend fun append(aggregateId: String, events: List<DomainEvent>)
    suspend fun load(aggregateId: String): List<DomainEvent>
    suspend fun loadAll(): List<DomainEvent>
}

class InMemoryEventStore : EventStore {
    private val store = ConcurrentHashMap<String, MutableList<DomainEvent>>()

    override suspend fun append(aggregateId: String, events: List<DomainEvent>) {
        store.computeIfAbsent(aggregateId) { mutableListOf() }.addAll(events)
    }

    override suspend fun load(aggregateId: String): List<DomainEvent> {
        return store[aggregateId]?.toList() ?: emptyList()
    }

    override suspend fun loadAll(): List<DomainEvent> {
        return store.values.flatMap { it.toList() }
    }
}
